### ISSUE TITLE
How to enable webgl  support?

### ISSUE THREAD

**Issue by lygstate** [Zenika/alpine-chrome#28]
Currently, there is no webgl support.( Softversion)

**Comment by jlandure** [comment]
Hi @lygstate 👋 

Thanks for your feedback.
I'm interested to help you. 

Can you describe a usecase with some code and the error you get?

Is it something like describe on this image [here](https://github.com/adieuadieu/serverless-chrome/issues/108)?

**Comment by lygstate** [comment]
I found that  only need pack swiftshader libegl and libglesv2

**Comment by jlandure** [comment]
Hi @lygstate 
Thanks for the update. 👍 
Do you think it could be useful for the community to get a `zenika/alpine-chrome:with-webgl` adding the missing packages ? Can you share a tiny example that works well?
Thanks for your contribution 🙌

**Comment by lygstate** [comment]
@jlandure  Need pacth the upstrem chromium packages:)
https://git.alpinelinux.org/aports/tree/community/chromium

But don't know where to start, patch file added
```
@@ -294,6 +269,11 @@ package() {
 	done
 	paxmark -m "$pkgdir"/usr/lib/$pkgname/chrome
 
+	install -Dm4755 libEGL.so "$pkgdir"/usr/lib/$pkgname/libEGL.so
+	install -Dm4755 libGLESv2.so "$pkgdir"/usr/lib/$pkgname/libGLESv2.so
+
+	install -Dm4755 swiftshader/libEGL.so "$pkgdir"/usr/lib/$pkgname/swiftshader/libEGL.so
+	install -Dm4755 swiftshader/libGLESv2.so "$pkgdir"/usr/lib/$pkgname/swiftshader/libGLESv2.so
 	install -Dm4755 chrome_sandbox "$pkgdir"/usr/lib/$pkgname/chrome-sandbox
 	install -m644 icudtl.dat "$pkgdir"/usr/lib/$pkgname/icudtl.dat
```

**Comment by maxcct** [comment]
I've also come here looking for the same thing! A `zenika/alpine-chrome:with-webgl` would be perfect and very much appreciated.

**Comment by jlandure** [comment]
Hi @maxcct @lygstate 👋 
I've tried several things today to add `webgl` support to the headless chrome.
Here is the PullRequest #29 

**But It's not working...**
I would like to get these outputs

`chromium-browser --headless --disable-dev-shm-usage --no-sandbox --screenshot --hide-scrollbars https://webglfundamentals.org/webgl/webgl-fundamentals.html`

Got

![webgl-fundamentals](https://user-images.githubusercontent.com/525974/59762650-fcc50c80-9297-11e9-86c5-735620138d06.png)

and expected is 

![webgl-fundamentals-expected](https://user-images.githubusercontent.com/525974/59762666-077fa180-9298-11e9-902d-e5172a25a756.png)

`chromium-browser --headless --disable-dev-shm-usage --no-sandbox --screenshot --hide-scrollbars https://browserleaks.com/webgl`

Got 

![webgl-activated](https://user-images.githubusercontent.com/525974/59762698-1fefbc00-9298-11e9-814a-20d426ade952.png)

and expected is 

![webgl-activated-expected](https://user-images.githubusercontent.com/525974/59762707-2847f700-9298-11e9-94bd-5c763be589c0.png)

Can you help me to make this `zenika/alpine-chrome:with-webgl` working? 👍

**Comment by lygstate** [comment]
@jlandure you need recompile chrome, use mesa gles not working, I've tried.

**Comment by jlandure** [comment]
Hi @lygstate 😊

Sorry for the delay.
I didn't achieve with success to recompile Chrome for the webgl support. 😣
Can you help me giving me the commands to recompile Chromium? 😇
Thanks!

**Comment by freedev** [comment]
I'm struggling with the same problem, any update on this?

**Comment by lygstate** [comment]
Sorry, wait a minute

**Comment by lygstate** [comment]
Change the APKBUILD file
```

package() {
	cd "$builddir"/out/$_buildtype
	local bin pak
	# paxmark inside chroot too
	paxmark -m mksnapshot

	# Installed files refere to
	# https://launchpad.net/ubuntu/disco/amd64/chromium-browser
	# https://dl.google.com/linux/direct/google-chrome-stable_current_amd64.deb
	# https://www.slimjet.com/chrome/google-chrome-old-version.php

	for bin in chrome chromedriver *.bin; do
		install -Dm755 $bin "$pkgdir"/usr/lib/$pkgname/$bin
	done
	paxmark -m "$pkgdir"/usr/lib/$pkgname/chrome

	install -Dm755 swiftshader/libEGL.so "$pkgdir"/usr/lib/$pkgname/swiftshader/libEGL.so
	install -Dm755 swiftshader/libGLESv2.so "$pkgdir"/usr/lib/$pkgname/swiftshader/libGLESv2.so

	install -Dm4755 chrome_sandbox "$pkgdir"/usr/lib/$pkgname/chrome-sandbox
	install -m644 icudtl.dat "$pkgdir"/usr/lib/$pkgname/icudtl.dat

	install -Dm755 xdg-mime "$pkgdir"/usr/lib/$pkgname/xdg-mime
	install -Dm755 xdg-settings "$pkgdir"/usr/lib/$pkgname/xdg-settings

	for pak in *.pak; do
		install -Dm644 $pak "$pkgdir"/usr/lib/$pkgname/$pak
	done

	install -Dm755 "$srcdir"/chromium-launcher.sh \
		"$pkgdir"/usr/lib/$pkgname/chromium-launcher.sh

	cp -a locales "$pkgdir"/usr/lib/$pkgname/
	cp -a MEIPreload "$pkgdir"/usr/lib/$pkgname/

	# It is important that we name the target "chromium-browser",
	# xdg-utils expect it; bug #355517.
	mkdir -p "$pkgdir"/usr/bin
	cd "$pkgdir"/usr/bin
	ln -sf /usr/lib/$pkgname/chromium-launcher.sh chromium-browser
	ln -sf /usr/lib/$pkgname/chromedriver "$pkgdir"/usr/bin/

	install -Dm644 "$srcdir"/chromium.conf \
		"$pkgdir"/etc/chromium/chromium.conf

	install -Dm644 "$srcdir"/chromium.desktop \
		"$pkgdir"/usr/share/applications/chromium.desktop

	cd "$builddir"
	for size in 24 48 64 128 256; do
		install -Dm644 "chrome/app/theme/chromium/product_logo_$size.png" \
			"$pkgdir/usr/share/icons/hicolor/${size}x${size}/apps/chromium.png"
	done

	for size in 16 32; do
		install -Dm644 "chrome/app/theme/default_100_percent/chromium/product_logo_$size.png" \
			"$pkgdir/usr/share/icons/hicolor/${size}x${size}/apps/chromium.png"
	done
}
```

**Comment by lygstate** [comment]
Basically,copy the file content of swiftshader
```
install -Dm755 swiftshader/libEGL.so "$pkgdir"/usr/lib/$pkgname/swiftshader/libEGL.so
install -Dm755 swiftshader/libGLESv2.so "$pkgdir"/usr/lib/$pkgname/swiftshader/libGLESv2.so
```

**Comment by pxlrbt** [comment]
@lygstate You got it working? Would you mind creating a PR with a working Dockerfile?

**Comment by jlandure** [comment]
@lygstate Please help us with this problem 👍
Open a PR or update the branch of the #29

**Comment by mlovatonv** [comment]
@lygstate you got it working? I copied the file content of swiftshader but it still doesn't work for me. How did you recompile chromium?

**Comment by lygstate** [comment]
Please refer to https://gitlab.alpinelinux.org/alpine/aports/-/merge_requests/6509

**Comment by pxlrbt** [comment]
@jlandure I'm currently building chromium as @lygstate suggested. Will create a PR if I succeed.

**Comment by jlandure** [comment]
Any news @lygstate @pxlrbt ? 👍

**Comment by pxlrbt** [comment]
@jlandure It's still building on my machine. And if it's successful I still need to integrate it into your Dockerfile. Somehow I could install `alpine-sdk` yesterday which is required.

**Comment by pxlrbt** [comment]
@jlandure Compilation is finally done and it worked with the example `chromium-browser --headless --disable-dev-shm-usage --no-sandbox --screenshot --hide-scrollbars https://webglfundamentals.org/webgl/webgl-fundamentals.html`.
I will check how to fix the issue with alpine-sdk after the weekend.

**Comment by pxlrbt** [comment]
@jlandure I got all issues before compilation step fixed within the Dockerfile. I am currently running a build on Docker Hub since last build took about 1,5 days on my machine. Maybe it's faster, maybe it will be aborted. Will create a PR if build is successfull.

**Comment by pxlrbt** [comment]
@jlandure I created a PR with a prebuilt and a build from source version here: https://github.com/Zenika/alpine-chrome/pull/50

**Comment by jlandure** [comment]
Hi @pxlrbt 

Thanks for your work! 🙌  How much times does it take to compile Chromium?
I will have a look.

FYI, I think there is a way to use [multi-stage build](https://docs.docker.com/develop/develop-images/multistage-build/) to create the binary and *NOT* storing in the repository 🤔

And... I have to create a `contributor` section to reward the contribution 👍 
Thanks again.

**Comment by pxlrbt** [comment]
@jlandure It took around 1.5 days to compile chromium on my machine. Probably should have a look at @lygstates solutions here https://github.com/Zenika/alpine-chrome/pull/51
I didn't test his result, but it looks way cleaner since he only add the missing libraries and there is  no need to recompile chromium again when new versions are released.

**Comment by jlandure** [comment]
@all-contributors please add @pxlrbt for question

**Comment by jlandure** [comment]
✅ This issue can be closed since this commit b5e7e1d0db27a1cfcd41d626e88f801747371c1d 🎉
